Remove redundant aggregation on username

gherceg opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi-dev.atlassian.net/browse/SAAS-12878

The `get_app_submission_breakdown_es` query takes too long. Seeing as aggregating on both user_id and username _should_ be redundant, I removed the username aggregation. We can still grab the username from the couch user and set it on the `MALTRow` object to not disrupt the MALT. I tested this in a shell and saw solid timing improvements, though this would still qualify as a slow query and is subject to change based on current load on ES. 

I removed the `_user_data` logic because after doing some investigation, I saw that:
- the most recent user_type of "AdminUser" was from November 2016
- the most recent user_type of "DemoUser" was from August 2017
- the most recent user_type of "UnknownUser" was also from August 2017

The only two user_types found in the MALT since September 2017 are "CommCareUser" and "WebUser" which come from the user doc_type field. Just in case, I added a `notify_exception` if we run into a situation where the user_id fetched is not part of the list of user_ids specified.

More concerning than removing those special user types, and the focus of this PR, is removing the username aggregation and potential risks involved with that. I sampled a random date (June 1st, 2021), and of the 99,226 records for that date, there wasn't a single case where the user fetched with the user_id had a different username than the username saved on the MALTRow. Even if we factor in that there is a possibility of deleting mobile workers and recycled usernames, that is still a reason to aggregate only on user_id and not username. 
## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
Malt generation is currently turned off on production, so this wouldn't have any immediate impacts.

Ideally, I'd initially create a limited release to do timing comparisons on production size domains.
### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Updated tests for the `get_app_submission_breakdown_es` query.
### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
No QA.
### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
remove redundant aggregation on username

czue reviewed (commented):
All makes sense, and code looks good to me! I'm just holding off on approval because it's been ages since I've been in this code and looks like there are plenty of other reviewers listed.

mjriley commented on corehq/apps/data_analytics/malt_generator.py:
```diff
@@ -63,25 +66,28 @@ def _get_malt_row_dicts(self, domain_name, monthspan, users_by_id):
             apps_submitted_for = get_app_submission_breakdown_es(domain_name, monthspan, users)
```
Not directly related, but it looks like the chunked call above is redundant here, as `users_by_id` was already sliced into chunks of 1000 within `build_table`.

gherceg replied:
Great point!

gherceg replied:
See 8aa83be048ca7b3adc8b8d7ff04d7c250eddf10a

mjriley commented on corehq/apps/data_analytics/tests/test_esaccessors.py:
```diff
@@ -31,25 +31,20 @@ def setUp(self):
+         ('app1', 'device', 'userid', 2),
      ],),
 ], MaltAnalyticsTest)
 def test_app_submission_breakdown(self, combination_count_list):
```
Was this the test suite where it was considered out-of-scope to add full tests? Notably, the code was modified to log when the user and user_id do not match, and that has no coverage here (I think?)

gherceg replied:
Since I've removed the code that checks/logs the user/user_id mismatch, with the explanation for doing so found [here](https://github.com/dimagi/commcare-hq/pull/31001#discussion_r793001256), this is no longer a concern.

mjriley reviewed (approved):
This seems fine, although I wouldn't object to more thorough tests

joxl commented on corehq/apps/data_analytics/malt_generator.py:
```diff
@@ -63,25 +66,28 @@ def _get_malt_row_dicts(self, domain_name, monthspan, users_by_id):
-                    )
+                    if app_row.user_id not in users_by_id:
+                        # highly unlikely, but adding a safety check for a recent refactor
+                        notify_exception(None, f"A form submission for user with id {user_id} was returned by "
+                                               f"get_app_submission_breakdown_es but was not found in the "
+                                               f"users_by_id parameter.")
```
What visibility do `notify_exception()` events receive? Is there a guaranteed investigation into the exception? If not, I might advocate a hard (uncaught) exception that fails this part of the report so the implications of this condition cannot go unaddressed by the consumers of the report.

gherceg replied:
I'm tempted to just remove that check entirely because it is essentially showing a mistrust in ElasticSearch results/filtering and the same level of awareness can be achieved by just attempting to access `users_by_id[app_row.user_id]`. Granted that would cause the entire task to fail if the id is does not exist, but that would actually provide better visibility in that we would see malt generation celery task failures in datadog. I could even setup an alert in datadog to fire if that task fails. 

If we wanted to stick with something closer to the current implementation, I believe there is a way to send an email for specific exceptions, but again my stance at the moment is that this code is unnecessary to continue maintaining. 

@joxl @mjriley let me know if you feel differently.

joxl replied:
> Granted that would cause the entire task to fail if the id is does not exist

:heavy_plus_sign:

joxl reviewed (commented):
Looks good, curious about the possibility of the edge-case condition going unnoticed and what that might mean in regards to the accuracy of the output.

gherceg committed:
rename _app_data to _get_malt_app_data and add tests

gherceg committed:
pull out _build_malt_row_dict method with tests, and add test for _get_malt_row_dicts

gherceg committed:
cleanup: run isort, log month info along with domain when buildilng malt

gherceg committed:
refactor MALTTableGenerator into a method: generate_malt

gherceg committed:
use django update_or_create instead of custom; tests for db save methods

gherceg commented:
@mjriley ask and you shall receive. Thanks for encouraging tests. I want to highlight that I removed the general `catch Exception` logic since it isn't good practice to catch broad exceptions, and am of the perspective that if we encounter specific exceptions in this code in the future, it is better to have the task fail and address those specific exceptions accordingly. **Definitely recommend reviewing the most recent push commit by commit**.

gherceg committed:
update comment

gherceg commented on corehq/apps/data_analytics/malt_generator.py:
```diff
@@ -52,8 +52,11 @@ def build_table(self, domains=None):
+                            f"Failed to get rows for domain {domain.name}. Exception is {str(ex)}", exc_info=True
+                        )
+                        continue
+
```
As mentioned below, I removed the general exception handling from this code.

mjriley reviewed (approved):
The changes look great, thanks for the tests. I'd like to see if we can steer away from using mocks as dummy objects, but beyond that, looks good!

gherceg committed:
use real app instead of mock in TestGetMaltAppData

gherceg committed:
remove explicit classes from super() call as it is no longer needed

gherceg committed:
use setUp instead of setUpClass whenever possible

gherceg committed:
specify the monthspan within the test to increase readability

gherceg committed:
patch method in setup/teardown where applicable

gherceg committed:
use web and mobile user instead of mock

gherceg committed:
remove redundant methods for creating app rows

gherceg committed:
do not use mock as variable name; move malt_row_dict create to setUp for TestUpdateOrCreateMaltRow

gherceg commented on corehq/apps/data_analytics/tests/test_malt_generator.py:
```diff
@@ -147,3 +151,41 @@ def test_missing_app_id_is_included(self):
+        self.assertEqual(actual_app_data, self.default_app_data)
+
+    def test_returns_expected_app_data(self):
+        mock_app = Mock()
```
Yeah I got a bit carried away and defaulted to using mocks. Thanks for the reminder. See 6f7fd4e7d1d8a1278908cbf82119faff36a680ec.

gherceg commented on corehq/apps/data_analytics/tests/test_malt_generator.py:
```diff
@@ -189,3 +189,138 @@ def test_returns_expected_app_data(self):
+
+    @classmethod
+    def setUpClass(cls):
+        super(TestBuildMaltRowDict, cls).setUpClass()
```
See f7f4097c389278a565fc466917fbf0e35e773ea6

gherceg commented on corehq/apps/data_analytics/tests/test_malt_generator.py:
```diff
@@ -189,3 +189,138 @@ def test_returns_expected_app_data(self):
+class TestBuildMaltRowDict(SimpleTestCase):
+
+    @classmethod
+    def setUpClass(cls):
```
See 3fd5a1a61908d32b5a995b163f0eec576de1df28

gherceg commented on corehq/apps/data_analytics/tests/test_malt_generator.py:
```diff
@@ -189,3 +189,138 @@ def test_returns_expected_app_data(self):
+
+    def test_returns_expected_value_for_month(self):
+        with patch('corehq.apps.data_analytics.malt_generator._get_malt_app_data', return_value=self.app_data):
+            actual_malt_row_dict = _build_malt_row_dict(self.app_row, self.domain, self.user, self.monthspan)
```
See 6dfb1874c458b6e6e1ee4c1959297aa8a71df148

gherceg commented on corehq/apps/data_analytics/tests/test_malt_generator.py:
```diff
@@ -189,3 +189,138 @@ def test_returns_expected_app_data(self):
+            DEFAULT_EXPERIENCED_THRESHOLD,
+            False
+        )
+        with patch('corehq.apps.data_analytics.malt_generator._get_malt_app_data', return_value=custom_app_data):
```
Yeah the namedtuple is immutable which means I can't replace `self.app_data` without also updating the return value for the patched method. See 8d529f3dbeed4ee38fabe05f80282e40d4e1cedf.

gherceg commented on corehq/apps/data_analytics/tests/test_malt_generator.py:
```diff
@@ -189,3 +189,138 @@ def test_returns_expected_app_data(self):
+
+
+def create_mock_user_for_malt_tests(user_id='user_id', username='username'):
+    mock_user = Mock()
```
See a9e01ba1eb536d23c31d904417a844034707cb45

gherceg committed:
no longer need to fetch domain objects, just names

gherceg committed:
only aggregate if there has been a form submission since the monthspan startdate

gherceg committed:
handle none returned for last submission date

gherceg commented:
Just noting that I add another optimization here, which is to not call `get_app_submission_breakdown_es` if there has not been a form submission since the start date of the current month. The `get_app_submission_breakdown_es` already returned no results if there weren't any submissions, so this is a safe optimization that has shown very promising results.

gherceg committed:
refactor MaltGeneratorTest to serve as one successful end to end test of malt generation

relying on lower level unit tests to handle specific cases like wam/pam values, missing app ids, etc

gherceg commented:
Made one last change to the top level MaltGeneratorTest to remove redundant test cases. I think I'm done 👍🏻

gherceg commented:
@mjriley @joxl could you look at the most recent changes one last time. Sorry for all the updates. I've been running this to rebuild the previous year's worth of malts with no issues. Would be great to get this into this week's deploy.

mjriley commented on corehq/apps/data_analytics/tests/test_malt_generator.py:
```diff
@@ -279,70 +273,51 @@ def setUp(self):
+        self.mock_get_malt_app_data = self.patcher.start()
+        self.mock_get_malt_app_data.return_value = self.app_data
+
+    def tearDown(self):
+        self.patcher.stop()
+        super().tearDown()
```
Take a look at the bottom example for patching [here](https://docs.python.org/3/library/unittest.mock.html#patch-methods-start-and-stop). It's less error-prone (and allows you to eliminate teardown entirely) if you use self.addCleanup(self.patcher.stop). Because the patcher is only used within the setup method, as well, it allows you to just refer to it as patcher, rather than self.patcher

gherceg replied:
See 80a39f63f862adc8bdea4bfc8cc3e1f3889082e1

mjriley commented on corehq/apps/data_analytics/tests/test_malt_generator.py:
```diff
@@ -348,29 +348,58 @@ def setUp(self) -> None:
         self.mock_get_malt_app_data = self.patcher.start()
         self.mock_get_malt_app_data.return_value = self.app_data
 
+        self.patcher2 = patch('corehq.apps.data_analytics.malt_generator.get_last_form_submission_received')
+        self.mock_get_last_form_submission = self.patcher2.start()
+        self.mock_get_last_form_submission.return_value = datetime.datetime(2022, 1, 10, 0, 0)
+
+        self.patcher3 = patch('corehq.apps.data_analytics.malt_generator.get_app_submission_breakdown_es')
```
Nit: could create better names for these patchers, i.e. `submissions_received_patcher` and `breakdown_es_patcher`. Not huge, but its a bit cryptic to see `patcher`1-3

gherceg replied:
See 80a39f63f862adc8bdea4bfc8cc3e1f3889082e1

mjriley reviewed (approved):
Looks great, thanks for the changes!

gherceg committed:
stop patchers via addCleanup instead of tearDown and more descriptive naming
